### PR TITLE
V2.1.0

### PR DESCRIPTION
**Pull request by SimonGLM**
# Changes

### DAEMON:
* Updates:
  - restructured data logging system: 
   - introduced log rotation and possibility to enable/disable data logging
   - simplified structure of data storage path under /var/muondetector
   - auto delete old data and log files
  - added geo position managing system with three modes of operation:
    - static: report position that was statically set (by user or as a result of a lock-in operation, see below)
    - auto: report position from most recent uBlox position fix
    - lock-in: fix a uBlox position only when accuracy is better than a user defined limit (for pos accuracy and pos DOP values) and make this the new static position
  - added several position filters applicable to auto and lock-in modes:
    - no filter: uBlox position fixes are not filtered (bad accuracy and bad precision))
    - kalman: apply a kalman filter which smoothes the series of last fixes efficiently but poorly filters long term transients/drifts (poor accuracy, excellent precision)
    - histogram mean (good accuracy, fair precision)
    - histogram median (better accuracy, fair precision)
    - histogram most probable value (mpv) (best accuracy, fair precision)
 
* Bugfixes:
  - removed wrong QT connect call which resulted in a warning in log messages after daemon startup
  - fixed reconnect to MQTT broker after connection was lost temporarily
  - fixed bug in I2C subsystem which lead to a stuck bus after certain read operations and preventing detection of a device in the next logical read access

### GUI
* Updates
  - Extended the geographic map feature with position marker (zoom level independent), semitransparent accuracy indicator and info box with current coordinates and position accuracy.
  - Added daemon hardware and software info under "Parameters" tab.
  - Added time as scan parameter in "Scans" tab.
  - Added display of log rotation interval and event log enable flag in "Log" tab.
  - Fixed bug which prevented auto update of histograms and plots.
  - Optimized color schemes to give better appearance for dark themes
  -  changed timing input selection from radio button group to drop down menu
  - Parameters: Added daemon hardware and software information
  - Scans: Added time as scan parameter
  - Log: Added display of log rotation interval and event log enable flag
  - Map: Intensively improved the geographic map
    - position marker (zoom level independent)
    - semitransparent accuracy indicator
    - info box with current coordinates and position accuracy
    - controls of geo position mode (auto, static, lock-in), static position and lock-in limits

* Bugfixes:
  - Fixed bug which prevented auto update of histograms and plots
  - Rate buffer was sometimes sent with incorrect size when connecting with the GUI leading to occasional crashes

### CODE REVIEW

File: daemon/include/calibration.h [hunk 1]
```diff
@@ -79,7 +77,8 @@ class ShowerDetectorCalib {
     void buildCalibList();
 
     std::vector<CalibStruct> fCalibList;
-    EEPROM24AA02* fEeprom = nullptr;
+    std::shared_ptr<EEPROM24AA02> fEeprom {};
+    //EEPROM24AA02* fEeprom = nullptr;
```
Comment (hangeza) [inline daemon/include/calibration.h]: delete commented-out code

File: daemon/include/daemon.h [hunk 2]
```diff
@@ -320,13 +223,10 @@ private slots:
     QString peerAddress;
     QHostAddress daemonAddress = QHostAddress::Null;
     quint16 peerPort, daemonPort;
-    QString gpsdevname;
     int verbose, baudrate;
     int gpsTimeout = 5000;
```
Comment (hangeza) [inline daemon/include/daemon.h]: move this constant to the other constant definitions

File: daemon/src/hardware/i2c/mic184.cpp [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+    if ((tos_save & 0x7f) != 0)
+        return false;
+    /*
+	std::cout << "MIC184::identify() : found LM75 base device at 0x"<<std::setw(2) << std::setfill('0')<<std::hex<<(int)fAddress<<"\n"; 
+	std::cout << " Regs: \n";
+	std::cout << "  conf  = 0x"<<std::setw(2) << std::setfill('0')<<(int)conf_reg_save<<"\n";
+	std::cout << "  thyst = 0x"<<std::setw(4) << std::setfill('0')<<thyst_save<<"\n";
+	std::cout << "  tos   = 0x"<<std::setw(4) << std::setfill('0')<<tos_save<<"\n";
+	std::cout << "  temp  = 0x"<<std::setw(4) << std::setfill('0')<<dataword<<"\n";
+*/
```
Comment (hangeza) [inline daemon/src/hardware/i2c/mic184.cpp]: remove code obviously intended for debugging purpose

File: daemon/src/hardware/i2c/pca9536.cpp [hunk 4]
```diff
@@ -39,5 +41,49 @@ bool PCA9536::devicePresent()
...
+        return false;
+    if (!devicePresent())
+        return false;
+    uint8_t bytereg { 0 };
+    /*	
+	for (int i=0; i<256; i++) {
+		if ( !readByte( static_cast<uint8_t>(i), &bytereg ) ) break;
+		std::cout << "reg 0x"<<std::hex<<std::setfill('0')<<std::setw(2)<<i<<" : 0x"<<(int)bytereg<<std::endl;
+	}
+*/
```
Comment (hangeza) [inline daemon/src/hardware/i2c/pca9536.cpp]: remove debugging code

File: daemon/src/hardware/i2c/pca9536.cpp [hunk 5]
```diff
@@ -39,5 +41,49 @@ bool PCA9536::devicePresent()
...
+    }
+    if ((bytereg & 0xf0) != 0xf0)
+        return false;
+    /*
+	if ( !readByte( static_cast<uint8_t>(REG::OUTPUT), &bytereg ) ) {
+		// there was an error
+		return false;
+	}
+	if ( ( bytereg & 0xf0 ) != 0xf0 ) return false;
+*/
```
Comment (hangeza) [inline daemon/src/hardware/i2c/pca9536.cpp]: remove commented-out code with no obvious functionality

File: gui/qml/places/items/MapComponent.qml [hunk 6]
```diff
@@ -90,7 +90,7 @@ Map {
         scaleText.text = text
     }
 
-    center: QtPositioning.coordinate(50.569212, 8.673828) // 2. Physikalisches Institut Gießen
+    center: QtPositioning.coordinate(0.569212, 8.673828) // 2. Physikalisches Institut Gießen
```
Comment (hangeza) [inline gui/qml/places/items/MapComponent.qml:93]: This assignment is obviously not true. Better to initialize the coordinates to "Zero Island" (lon=0, lat=0) and remove the comment indicating a different default location.

File: daemon/config/changelog [hunk 7]
```diff
@@ -1,3 +1,23 @@
+muondetector-daemon (2.1.0) ; urgency=low
```
Comment (hangeza) [inline daemon/config/changelog:1]: list more of the changes here, i.e. network discovery

File: daemon/src/main.cpp [hunk 8]
```diff
@@ -450,81 +428,82 @@ int main(int argc, char* argv[])
...
+    default:
+        daemonConfig.eventTrigger = EVT_AND;
+        break;
+    }
+
     if (parser.isSet(pol1Option)) {
         unsigned int pol1int = parser.value(pol1Option).toUInt(&ok);
         if (!ok || pol1int > 1) {
-            qCritical() << "wrong input polarity setting ch1 (valid: 0,1)";
+            qCritical() << "wrong input polarity setting ch1 (valid: 0..3)";
```
Comment (hangeza) [inline daemon/src/main.cpp]: something wrong here: The program exits, when another value than 0 or 1 is read for the polarity (yet, the error message indicates a valid range btw. 0 and 3). Instead, the program should set a sane default value in this case.

File: gui/src/mainwindow.cpp [hunk 9]
```diff
@@ -608,103 +612,101 @@ void MainWindow::receivedTcpMessage(TcpMessage tcpMessage)
...
         emit mqttInhibitReceived(inhibit);
-        updateUiProperties();
+        return;
+    } else if (msgID == TCP_MSG_KEY::MSG_VERSION) {
+        MuonPi::Version::Version hw_ver, sw_ver;
+        *(tcpMessage.dStream) >> hw_ver >> sw_ver;
+        emit daemonVersionReceived(hw_ver, sw_ver);
+        return;
+    } else {
+        qDebug() << "received unknown TCP message, msgID =" << QString::number(static_cast<int>(msgID));
```
Comment (hangeza) [inline gui/src/mainwindow.cpp]: this should rather be a warning, shouldn't it?
```suggestion
        qWarning() << "received unknown TCP message, msgID =" << QString::number(static_cast<int>(msgID));
```

**Review by hangeza — COMMENTED** [review]
Some minor issues still to be changed and changelogs should be updated.

**Comment by hangeza** [comment]
Be aware that the merge of the branch position-models into dev with PR #83 was reverted. Hence, the position-models functionality is NOT part of dev, v2.1.0 and specifically this PR.
Should we wait with this PR until the position models is completed and finally merged with dev?

**Comment by SimonGLM** [comment]
@hangeza You're right! We should wait until that feature is ready to be merged.

**Pull request merged**